### PR TITLE
refactor: Rust filesystem modules

### PR DESCRIPTION
**Pull request by dubiousjim**
This is part of a series of smaller PRs towards #4017.

It improves debug! calls, and does some other cleanup-type refactoring in Rust filesystem modules.

### CODE REVIEW

File: cli/ops/fs.rs [hunk 1]
```diff
@@ -6,7 +6,6 @@ use crate::op_error::OpError;
 use crate::ops::dispatch_json::JsonResult;
 use crate::state::State;
 use deno_core::*;
-use remove_dir_all::remove_dir_all;
```
Comment (bartlomieju) [inline cli/ops/fs.rs]: I think this import can be removed (along with entry in `Cargo.toml`) and replaced with [`std::fs::remove_dir_all`](https://doc.rust-lang.org/std/fs/fn.remove_dir_all.html)

File: cli/ops/fs.rs [hunk 2]
```diff
@@ -611,7 +618,7 @@ fn op_utime(
   let is_sync = args.promise_id.is_none();
   blocking_json(is_sync, move || {
     debug!("op_utime {} {} {}", args.filename, args.atime, args.mtime);
-    utime::set_file_times(args.filename, args.atime, args.mtime)?;
+    deno_fs::set_file_times(args.filename, args.atime, args.mtime)?;
```
Comment (bartlomieju) [inline cli/ops/fs.rs]: Unnecessary change of import

File: cli/ops/fs.rs (hunk 1 above)
Comment (dubiousjim) [reply cli/ops/fs.rs]: Will do. Probably this is a good opportunity to start shifting over to the tokio::fs versions, as you propose in #4188.

Q: after removing the import from `Cargo.toml`, should I also commit an updated `Cargo.lock`? Or do you defer that until just before a release?

Comment (bartlomieju) [reply cli/ops/fs.rs]: Yes, but #4188 should be done in a separate PR (I think I saw this code in one of your PRs)

Yes, Cargo.lock should be committed as well

File: cli/ops/fs.rs (hunk 2 above)
Comment (dubiousjim) [reply cli/ops/fs.rs]: Let me explain the thinking that prompted this, it'd help to see what mistake I was making. Initially, I was confused what was the point of many operations being split between `cli/ops/fs.rs` and `cli/fs.rs`. For example, why is the function `chown` in `cli/fs.rs` rather than just being folded into the `op_chown` in `cli/ops/fs.rs`. (It's not invoked from anywhere else.) Similarly for the `mkdir` and `make_temp` code in `cli/fs.rs`.

I would have asked someone if I knew what was a good channel to do that in, but I didn't. Maybe the gitter lobby is the place to do that, but there's a lot of chatter there.

Anyway, as I was working on adding more filesystem ops, and expanding the functionality of some existing ones, I gradually got the impression that oh, the point of `cli/fs.rs` is to consolidate all of the call-outs to third-party crates (including `nix`, but not `tokio`), and maybe also the call-outs to `unix::fs` stuff. The existing codebase seemed to match that impression, but only 95%. There were still these two calls to `remove_dir_all` and `set_file_times` in `cli/ops/fs.rs`. So I thought, Ok, let's just make sure they're exported from `cli/fs.rs` (as `pub use`s), and have `cli/ops/fs.rs` import them from there. Hence this commit.

I agree that looked at from another angle, this change is pointless. But then I'd value some guidance in better understanding when new functionality should be partly placed in `cli/fs.rs`, as you guys did for `chown` etc, and what should be wholly included in `cli/ops/fs.rs`.

Comment (ry) [reply cli/ops/fs.rs]: > For example, why is the function chown in cli/fs.rs rather than just being folded into the op_chown in cli/ops/fs.rs.

Functionality should be folded into the ops themselves unless it's being used elsewhere. We have gone through many refactors, s things might be a certain way just from oversight rather than by design. I think that's the situation with chown.

Comment (dubiousjim) [reply cli/ops/fs.rs]: OK, I'll undo the attempts to consolidate the third-party imports into `cli/fs.rs`. For the moment I'll leave `chown` where it is --- and maybe I'll push some of the other stuff I've got in this sequence, which as it stands puts third-party stuff into `cli/fs.rs`. But at some point soon, once more of this stuff has worked its way into the codebase, I'll move `chown` etc (and anything I add to `cli/fs.rs`) back into `cli/ops/fs.rs`.

File: cli/ops/fs.rs (hunk 1 above)
Comment (dubiousjim) [reply cli/ops/fs.rs]: Ok, there's a problem using `std::fs::remove_dir_all` as I do in `97c1a69` (below). Then the tests fail on Windows with a `error: Uncaught Error: The directory is not empty. (os error 145)`. I assume this comes from `removeAllSyncDirSuccess` or `removeAllDirSuccess`, but those tests aren't currently written so as to catch that failure. So for the moment, I guess we should stick with `remove_dir_all::remove_dir_all`? It may be that the `tokio::fs` version works better on Windows.

Comment (bartlomieju) [reply cli/ops/fs.rs]: Okay, makes sense 👍

Comment (ry) [reply cli/ops/fs.rs]: @dubiousjim I don't understand - these tests currently are fine?

I really don't want to add new crate dependencies for functionality that is in std.

File: cli/fs.rs [hunk 3]
```diff
@@ -95,7 +96,7 @@ pub fn make_temp(
 }
 
 pub fn mkdir(path: &Path, perm: u32, recursive: bool) -> std::io::Result<()> {
-  debug!("mkdir -p {}", path.display());
+  // debug!("mkdir -p {}", path.display());
```
Comment (ry) [inline cli/fs.rs]: Unnecessary. You can either remove the line or leave it as is, but leaving dead code isn't helpful. Suggest leaving it as is.

File: cli/js/remove_test.ts [hunk 4]
```diff
@@ -169,7 +169,13 @@ testPerm({ write: true, read: true }, function removeAllSyncDirSuccess(): void {
...
   const subPathInfo = Deno.statSync(subPath);
   assert(subPathInfo.isDirectory()); // check exist first
-  Deno.removeSync(path, { recursive: true }); // remove
+  let success = true;
+  try {
+    Deno.removeSync(path, { recursive: true }); // remove
+  } catch (e) {
+    success = false;
+  }
+  assert(success);
```
Comment (ry) [inline cli/js/remove_test.ts]: What's the purpose of this try-catch if it's supposed to succeed?

File: cli/js/remove_test.ts [hunk 5]
```diff
@@ -403,7 +409,13 @@ testPerm(
...
     const subPathInfo = Deno.statSync(subPath);
     assert(subPathInfo.isDirectory()); // check exist first
-    await Deno.remove(path, { recursive: true }); // remove
+    let success = true;
+    try {
+      await Deno.remove(path, { recursive: true }); // remove
+    } catch (e) {
+      success = false;
+    }
+    assert(success);
```
Comment (ry) [inline cli/js/remove_test.ts]: ditto

**Comment by bartlomieju** [comment]
@dubiousjim heads up: I've made significant changes to test, could you resolve the conflicts?

File: cli/js/remove_test.ts (hunk 4 above)
Comment (dubiousjim) [reply cli/js/remove_test.ts]: Yeah, this PR has gotten confusing as it's evolved. It might be better for me to just close it and re-push a new version of it with some of these commits squashed together. But for explanation, here's what's going on/how we got here. Other than the `debug!` improvements, there's not much else in this PR except a bit of refactoring. One thing I did with the refactoring was move the third-party imports to `cli/fs.rs`, because I thought that was the guiding intent about how to divide work between `cli/fs.rs` and `cli/ops/fs.rs`. You explained in an earlier comment thread on this PR that no, things are probably just divided as they are because of the contingencies of past refactorings, and stuff should ideally all be in `cli/ops/fs.rs` if it can be. So I pushed revisions to this branch that moved things back. Also, @bartlomieju noticed that one of the imports I was moving back and forth between the files was of the third-party crate `remove_dir_all`. He said, we don't need that, there's such a function in `std::fs`. I hadn't noticed that, so I tried replacing the use of `remove_dir_all::remove_dir_all` with a plain `std::fs::remove_dir_all`. That turned out to work fine on Mac and Unix, but when I pushed it to the branch this PR is based on, the CI test for Windows turned out to fail. (See above comment exchange between me and @bartlomieju.) So I pushed another revision that re-instated the use of `remove_dir_all::remove_dir_all`, at least for the time being until we worked out a good solution for all platforms. Also, I noticed that the report of the failure on Windows was funny: it didn't say that a specific test failed, but there was just a `error: Uncaught Error: The directory is not empty. (os error 145).` I don't understand why. I tried adding the `try ... catch` blocks to those tests, so that if someone else breaks this in the same way later, they'd get a better error report. But maybe those fixes are stupid, and would fail in just the same way that the bare `remove` call did.

Anyway, hope that explains what's going on. Should I close this PR and resubmit a squashed/pruned version of it?

**Comment by dubiousjim** [comment]
> @dubiousjim heads up: I've made significant changes to test, could you resolve the conflicts?

Done.

File: cli/js/remove_test.ts (hunk 4 above)
Comment (dubiousjim) [reply cli/js/remove_test.ts]: @ry, check out [this branch on my fork](https://github.com/dubiousjim/deno/commits/filesystem1bb). That has six commits on top of denoland/master, the first four being the good parts of this PR, and the last two replacing the use of `remove_dir_all::remove_dir_all` with `std::fs::remove_dir_all`. Based on what happened before, I'm expecting the branch to fail CI tests on Windows, and so the last two commits should be reverted/dropped. That would be the pruned version of this PR.

Comment (dubiousjim) [reply cli/js/remove_test.ts]: Confirmed that that fails, with only this information:

> error: Uncaught Error: The directory is not empty. (os error 145)
> $deno$/dispatch_json.ts:41:11
>     at unwrapResponse ($deno$/dispatch_json.ts:41:11)
>     at sendAsync ($deno$/dispatch_json.ts:96:10)
> test tests::std_tests ... FAILED
> 
> failures:
> 
> ---- tests::std_tests stdout ----
> target_dir D:\a\deno\deno\target\release
> ##[error]thread 'tests::std_tests' panicked at 'assertion failed: status.success()', cli\tests\std_tests.rs:31:5
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
> error: test failed, to rerun pass '-p deno --test std_tests'
> 
> failures:
>     tests::std_tests
> 
> test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
> 
> ##[error]Process completed with exit code 1.

Comment (dubiousjim) [reply cli/js/remove_test.ts]: Fails only on windows.

**Comment by dubiousjim** [comment]
Am going to `push --force` a new branch to my fork overwriting this one, that contains just a pruned version of the useful parts of this PR, without the confusing history. Not sure how it will come out on this page. If it's a mess, I'll close this PR and open a new one.

**Comment by dubiousjim** [comment]
Hm, this page shows the `push --force` but doesn't present the commits for review. Guess I should close & reopen a different PR.

**Comment by ry** [comment]
@dubiousjim I think it's better to combine these changes (if at all) with some other ones. Closing.
